### PR TITLE
317 Refactor Observability Stack with OpenTelemetry

### PR DESCRIPTION
**Pull request by ginaxu1**
## Summary

This PR connects Go services to the observability stack by implementing **OpenTelemetry-based metrics**. This enables Prometheus to scrape metrics from these services and display them in Grafana dashboards. Services now use vendor-agnostic OpenTelemetry instrumentation to allow seamless switching between Prometheus (default for local dev), Datadog, New Relic, or any OTLP-compatible backend **without changing code** - just environment variables.

All services now expose the following Prometheus metrics:
- **`http_requests_total{http_method, http_route, http_status_code}`** - Total HTTP request count by method, route, and status code
- **`http_request_duration_seconds{http_method, http_route}`** - HTTP request latency histogram by method and route
- **`external_calls_total{external_target, external_operation}`** - External service call metrics (exchange services)
- **`business_events_total{business_action, business_outcome}`** - Business event metrics (exchange services)

**Why these changes are needed:**
- The observability stack (Prometheus + Grafana) is configured but cannot collect data without service instrumentation
- OpenTelemetry provides vendor-agnostic instrumentation, allowing teams to choose their observability backend (Prometheus, Datadog, New Relic, etc.) without code changes

## Type of Change
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [x] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Other (please describe):

### Architectural Changes

- **OpenTelemetry Integration**: All services now use OpenTelemetry SDK for vendor-agnostic metrics collection
- **Shared Monitoring Package**: Created reusable monitoring package in `exchange/shared/monitoring/` for all exchange services
- **Backward-Compatible API**: Old API (`monitoring.Handler()`, `monitoring.HTTPMetricsMiddleware()`) still works, delegates to OpenTelemetry
- **Auto-Initialization**: Metrics initialize automatically when first used (no explicit initialization needed)
- **Non-Breaking Integration**: Metrics are added without modifying existing handler logic
- **Vendor-Agnostic**: Switch between Prometheus, Datadog, New Relic via environment variables (no code changes)

## Testing

- [x] I have tested this change locally
- [x] I have added unit tests for new functionality
- [x] I have tested edge cases
- [x] All existing tests pass

### Runtime Testing

To verify the observability stack is working:

1. **Start observability stack:**
   ```bash
   cd observability
   ./start-grafana.sh  # or: docker compose up -d
   ```
2. **Check Prometheus targets:**
   - Open http://localhost:9091/targets
   - Services should show as "UP" (green)

3. **Generate sample traffic:**
   ```bash
   cd observability
   ./generate_sample_traffic.sh
   ```

4. **View metrics in Grafana:**
   - Open http://localhost:3002/d/go-services/go-services-metrics
   - Metrics should appear after services receive traffic

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have checked that there are no merge conflicts
- [x] I have verified all services are on `opendif-network`
- [x] I have verified Prometheus can scrape all service endpoints

### Pre-Deployment Checklist

1. **Service Restart Required**: Services must be restarted to load the new monitoring code
   ```bash
   # Stop existing services
   # Rebuild: go build .
   # Restart services
   ```

2. **Network Setup**: Ensure `opendif-network` exists before starting services
   ```bash
   docker network create opendif-network  # if it doesn't exist
   # Or use: cd observability && ./start-grafana.sh
   ```

3. **No Configuration Changes Required**: Services use Prometheus exporter by default (no env vars needed for local dev)

4. **Prometheus Already Configured**: Prometheus is already configured to scrape these services (see `observability/prometheus/prometheus.yml`)

5. **Grafana Dashboard Ready**: Grafana dashboard is already configured to display these metrics

### Environment Variables (Optional)

For local development, no environment variables are needed (Prometheus is default).

To switch to other backends (Datadog, New Relic, etc.), set:
```bash
export OTEL_METRICS_EXPORTER=otlp
export OTEL_EXPORTER_OTLP_ENDPOINT=<your-endpoint>
export OTEL_EXPORTER_OTLP_HEADERS="<your-headers>"
export SERVICE_NAME=<service-name>
```

### Post-Deployment Verification

1. **Check Metrics Endpoints**:
   ```bash
   curl http://localhost:4000/metrics | grep http_requests_total
   curl http://localhost:8082/metrics | grep http_requests_total
   curl http://localhost:3000/metrics | grep http_requests_total
   ```

2. **Verify Prometheus Scraping**:
   - Open http://localhost:9091/targets
   - All services should show as "UP" (green)

3. **View in Grafana**:
   - Open http://localhost:3002/d/go-services/go-services-metrics
   - Metrics should appear after services receive some traffic

4. **Generate Sample Traffic**:
   ```bash
   cd observability
   ./generate_sample_traffic.sh
   ```

## Migration Notes

- **Backward Compatible**: The API remains the same - existing code continues to work
- **Auto-initialization**: Metrics initialize automatically when first used
- **No code changes required**: Services using `monitoring.Handler()` or `monitoring.HTTPMetricsMiddleware()` work without changes
- **OpenTelemetry Under the Hood**: Prometheus client is now indirect dependency via OpenTelemetry Prometheus exporter

## Future Work

- Add metrics instrumentation to Consent Engine and Audit Service
- Add custom business metrics for specific use cases
- Configure alerting rules in Prometheus
- Add distributed tracing (OpenTelemetry traces)

### CODE REVIEW

File: exchange/shared/monitoring/otel_metrics.go [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+	return initErr
+}
+
+// initializeInternal performs the actual initialization work
+func initializeInternal(ctx context.Context, config Config) error {
+	// Create resource with service name
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			semconv.ServiceName(config.ServiceName),
+			semconv.ServiceVersion("dev"),
```
Comment (mushrafmim) [inline exchange/shared/monitoring/otel_metrics.go]: Why "dev" is hardcoded?

File: exchange/shared/monitoring/metrics.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+// (not normalized) to prevent cardinality explosion in metrics.
+// This map is initialized once at package load time and reused for all requests.
+var knownRoutes = map[string]bool{
+	"/health":                         true,
+	"/metrics":                        true,
+	"/debug":                          true,
+	"/api/v1/policy/metadata":         true,
+	"/api/v1/policy/decide":           true,
+	"/api/v1/policy/update-allowlist": true,
+}
```
Comment (mushrafmim) [inline exchange/shared/monitoring/metrics.go]: knownRoutes should not be defined here. Since, it is a common package, these routes should be configured. Better to plug them in from each service.

**Review by mushrafmim — CHANGES_REQUESTED** [review]
Other than that this is looking good, since the `consent-engine` is added, I will validate whether it is working along properly and approve the PR.

**Review by mushrafmim — CHANGES_REQUESTED** [review]
